sdk/src/azure/core/az_span_private.h

```diff
@@ -63,7 +63,9 @@ AZ_NODISCARD az_result _az_span_replace(
     int32_t current_size,
     int32_t start,
     int32_t end,
-    az_span replacement);
+    az_span replacement,
+    bool url_encode,
+    int32_t* out_url_encode_size);
```

Comment:
I'm struggling with this because of the Single Responsibility Principle. It looks like this function is doing replace AND encoding - which is 2 responsibilities. Can we improve upon this design?

Reply:
I know.
The benefit of this is that this function is already taking care of calculating the required size and it  already knows how to do shift left/right on a buffer (what we need to append path).

Reply:
I agree with Jeff. We should try to write these methods independently so replace has nothing to do with encoding.

Reply:
replace is a private method and it is only used by append_path and set_query_parameters...

Would you prefer a new private method like `replace_with_url_encode` ?

Let's consider `append_path`.  If we need to url-encode and az_span before appending it to URL, and if the URL already contains query parameters, we would need to do a shift right.
-  We don't have an extra buffer to convert the az_span to url-encode before appending, we need to use URL

I understand this replace function might be doing too much (insert, remove, append, concatenate...)  and now it can also make a transformation to source az_span before replacing.
I think that, if in the future there is another transformation required (other than url-encode), we can create a type transformation with a transform_fn() and calc_siz_after_transform() than replace can call before replacing.